voting/assembly/main.ts

```diff
@@ -0,0 +1,119 @@
+import { context, 
+  env, 
+  PersistentUnorderedMap, 
+  u128 } from 'near-sdk-as'
+import { 
+  AccountId, 
+  Balance, 
+  EpochHeight, 
+  Result } from './types'
+import { 
+    ERR_VOTING_ALREADY_ENDED, 
+    ERR_CHECK_RESULT_AFTER_RESULT } from './non-spec'
+
+@nearBindgen
+export class VotingContract {
```

Comment:
Sorry for the late reply.

I thought I had addressed this a while ago in a comment (my bad!!).  The current issue is that you can't export a class in AS the same way as rust.

I've wanted to add this feature, but currently if you export a class it mangles the names of the methods (this is useful in a JS ecosystem where then you can create a JS Object with them). So what you have to do is something like


```ts
class VotingContract {
    // Only instance state and/or constructor
    this.votes = new PersistentUnorderedMap<AccountId, Balance>("v")
    this.total_voted_stake = new u128()
    this.result =  'None'
    this.last_epoch_height = 0 
}

let contract = storage.get<VotingContract>("STATE")

@exportAs("new")
export function create() {
  assert(contract == null, "contract already initialized);
  contract = new VotingContract();
  storage.set("STATE", contract);
}

export function check_result(): void {
  assert(contract != null, "contract should be initialized");
  let self = <VotingContract>contract; // So compiler knows it is non-null;
  assert(self.result == 'None', ERR_CHECK_RESULT_AFTER_RESULT);
   ...
  storage.set("STATE", self);
}

```

Feel free to reach out on discord if you need clarification.  This boilerplate could be solved during the bindgen phase, which is a bit todo!  Sorry again for the lack of communication and clarity!

Reply:
@willemneal can you point me to your discord name?